rust/src/composepost.rs

```diff
@@ -697,6 +697,70 @@ fn convert_path_to_tmpfiles_d_recurse(
     Ok(())
 }
 
+/// Walk over the root filesystem and perform some core conversions
+/// from RPM conventions to OSTree conventions.
+///
+/// For example:
+///  - Symlink /usr/local -> /var/usrlocal
+///  - Symlink /var/lib/alternatives -> /usr/lib/alternatives
+///  - Symlink /var/lib/vagrant -> /usr/lib/vagrant
+#[context("Preparing symlinks in rootfs")]
+pub fn rootfs_prepare_links(rootfs_dfd: i32) -> CxxResult<()> {
+    let rootfs = crate::ffiutil::ffi_view_openat_dir(rootfs_dfd);
+
+    rootfs
+        .remove_all("usr/local")
+        .context("Removing /usr/local")?;
+    let state_paths = &["usr/lib/alternatives", "usr/lib/vagrant"];
+    for entry in state_paths {
+        rootfs
+            .ensure_dir_all(*entry, 0o0755)
+            .with_context(|| format!("Creating '/{}'", entry))?;
+    }
+
+    let symlinks = &[
+        ("../var/usrlocal", "usr/local"),
+        ("../../usr/lib/alternatives", "var/lib/alternatives"),
+        ("../../usr/lib/vagrant", "var/lib/vagrant"),
+    ];
+    for (target, linkpath) in symlinks {
+        ensure_symlink(&rootfs, target, linkpath)?;
+    }
+
+    Ok(())
+}
+
+/// Create a symlink at `linkpath` if it does not exist, pointing to `target`.
+///
+/// This is idempotent and does not alter any content already existing at `linkpath`.
+/// It returns `true` if the symlink has been created, `false` otherwise.
+#[context("Symlinking '/{}' to empty directory '/{}'", linkpath, target)]
+fn ensure_symlink(rootfs: &openat::Dir, target: &str, linkpath: &str) -> Result<bool> {
+    use openat::SimpleType;
+
+    if let Some(meta) = rootfs.metadata_optional(linkpath)? {
+        match meta.simple_type() {
+            SimpleType::Symlink => {
+                // We assume linkpath already points to the correct target,
+                // thus this short-circuits in an idempotent way.
+                return Ok(false);
+            }
+            SimpleType::Dir => rootfs.remove_dir(linkpath)?,
+            _ => bail!("Content already exists at link path"),
+        };
+    } else {
+        // For maximum compatibility, create parent directories too.  This
+        // is necessary when we're doing layering on top of a base commit,
+        // and the /var will be empty.  We should probably consider running
+        // systemd-tmpfiles to setup the temporary /var.
+        rootfs.ensure_dir_all(linkpath, 0o755)?;
+        rootfs.remove_dir(linkpath)?;
```

Comment:
```suggestion
        rootfs.ensure_dir_all(Path::new(linkpath).parent().unwrap(), 0o755)?;
```
or so?

Reply:
Ah yes, that's actually closer to the original code. An `if let Some(parent)` would avoid the unwrap too.

Reply:
Pushed a commit on top, optimizing this logic in the suggested direction.